File: 0746-min-cost-climbing-stairs/0746-min-cost-climbing-stairs.java
class Solution {
    int[] cost;
    int[] memo;
    public int minCostClimbingStairs(int[] cost) {
        this.cost = cost;
        this.memo = new int[cost.length+1];
       
        return dfs(cost.length);
    }
    
    private int dfs(int i) {
        if (i < 2) {
            return 0;
        }
        if (memo[i] == 0) {
            memo[i] = Math.min(dfs(i-1)+cost[i-1], dfs(i-2)+cost[i-2]);
        }
        return memo[i];
    }
}

File: 0746-min-cost-climbing-stairs/0746-min-cost-climbing-stairs.cpp
class Solution {
public:
    int minCostClimbingStairs(vector<int>& cost) {
        unordered_map<int, int> memo;
        for (int i = 2; i <= cost.size(); i++) {
            memo[i] = min(memo[i-1]+cost[i-1], memo[i-2]+cost[i-2]);
        }
        return memo[cost.size()];
    }
};